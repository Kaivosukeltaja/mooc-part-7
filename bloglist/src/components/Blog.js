import React from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { addLike } from '../actions/blogs'

const blogStyle = {
  backgroundColor: '#ffffff',
  padding: '1em',
  borderRadius: '4px',
  color: '#333333',
  margin: '0.25em 0',
}

const blogHeadingStyle = {
  fontWeight: 'bold',
  fontSize: '1.25em',
  cursor: 'pointer',
}

const visibleDetails = {
  display: 'block',
}

const hiddenDetails = {
  display: 'none',
}

class Blog extends React.Component {
  constructor(props) {
    super(props)
    this.state = {
      showDetails: false,
    }
  }

  addLike = () => { this.props.addLike(this.props.blog) }

  toggleDetails = () => {
    this.setState({ showDetails: !this.state.showDetails })
  }

  deleteBlog = () => { this.props.deleteBlog(this.props.blog) }
  
  render() {
    return (
      <div className="blog" style={blogStyle}>
        <div
          className="blog__title"
          style={blogHeadingStyle}
          onClick={this.toggleDetails}  
        >
          {this.props.blog.title} {this.props.blog.author}
        </div>
        <div className="blog__details" 
          style={this.state.showDetails ? visibleDetails : hiddenDetails}
        >
          <div><a href={this.props.blog.url}>{this.props.blog.url}</a></div>
          { this.props.blog.user && 
            <div>Added by { this.props.blog.user.name }</div>
          }
          <div>Likes: {this.props.blog.likes} <button onClick={this.addLike}>Like</button></div>
          { this.props.deleteBlog &&
            <button onClick={this.deleteBlog}>Delete</button>
          }
        </div>
      </div>
    )
  }
}

export const blogShape = PropTypes.shape({
  title: PropTypes.string.isRequired,
  author: PropTypes.string.isRequired,
  likes: PropTypes.number.isRequired,
  url: PropTypes.string.isRequired,
  user: PropTypes.shape({
    name: PropTypes.string.isRequired,
    username: PropTypes.string.isRequired,
  })
})

Blog.propTypes = {
  addLike: PropTypes.func.isRequired,
  deleteBlog: PropTypes.func,
  blog: blogShape,
}

const mapDispatchToProps = {
  addLike
}

export default connect(null, mapDispatchToProps)(Blog)
