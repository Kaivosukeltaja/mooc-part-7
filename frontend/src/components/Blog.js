import React from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { Link } from 'react-router-dom'

const blogStyle = {
  backgroundColor: '#ffffff',
  padding: '1em',
  borderRadius: '4px',
  color: '#333333',
  margin: '0.25em 0',
}

const blogHeadingStyle = {
  fontWeight: 'bold',
  fontSize: '1.25em',
  cursor: 'pointer',
}

const Blog = (props) => (
  <div style={blogStyle}>
    <Link
      style={blogHeadingStyle}
      to={`/blogs/${props.blog._id}`}
    >
      {props.blog.title} - {props.blog.author}
    </Link>
  </div>
)

export const blogShape = PropTypes.shape({
  title: PropTypes.string.isRequired,
  author: PropTypes.string.isRequired,
  likes: PropTypes.number.isRequired,
  url: PropTypes.string.isRequired,
  user: PropTypes.shape({
    name: PropTypes.string.isRequired,
    username: PropTypes.string.isRequired,
  })
})

Blog.propTypes = {
  currentUser: PropTypes.shape(),
  blog: blogShape,
}

const mapStateToProps = (state) => ({
  currentUser: state.users.currentUser,
})

export default connect(mapStateToProps)(Blog)
