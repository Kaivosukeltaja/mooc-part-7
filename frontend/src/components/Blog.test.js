import React from 'react'
// import { shallow } from 'enzyme'
import { shallowWithStore } from 'enzyme-redux'
import { createMockStore } from 'redux-test-utils'
import renderer from 'react-test-renderer'
import MockRouter from 'react-mock-router'
import { Link } from 'react-router-dom'
import { Blog } from './Blog'

const mockState = {
  users: {
    currentUser: null,
  }
}

describe('<Blog />', () => {
  const blog = {
    title: 'Test blog',
    author: 'Test Author',
    url: 'https://google.com',
    likes: 123,
    user: {
      username: 'foobar',
      name: 'Foo Bar',
    }
  }

  const mockStore = createMockStore(mockState)

  it('renders correctly', () => {
    const tree = renderer
      .create(<MockRouter>
        <Blog blog={blog} />
      </MockRouter>)
      .toJSON()
    expect(tree).toMatchSnapshot()

  })

  it('contains a link to the article', () => {
    const blogComponent = shallowWithStore(<Blog blog={blog} />, mockStore)
    
    const titleLinks = blogComponent.find(Link)

    expect(titleLinks.length).toBe(1)
  })
})
