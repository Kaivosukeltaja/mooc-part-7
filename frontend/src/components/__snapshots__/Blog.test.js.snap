// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Blog /> renders correctly 1`] = `
<div
  style={
    Object {
      "backgroundColor": "#ffffff",
      "borderRadius": "4px",
      "color": "#333333",
      "margin": "0.25em 0",
      "padding": "1em",
    }
  }
>
  <a
    href={undefined}
    onClick={[Function]}
    style={
      Object {
        "cursor": "pointer",
        "fontSize": "1.25em",
        "fontWeight": "bold",
      }
    }
  >
    Test blog
     - 
    Test Author
  </a>
</div>
`;
